### ISSUE TITLE
Integrating touch into the HTML5 backend

### ISSUE THREAD

**Issue by alexandercrichton** [alexandercrichton/react-dnd-html5-backend#1]
Hi @gaearon and @longlho 

I'm sure you guys are awfully busy. I was just wondering if anyone else would find this combination of backends useful? I'm using it because I don't want to rely on touch detection to set up my backend. I'm supporting mouse and touch for any user. The two distinct backends handle their responsibilities very well, and it seems they can operate side-by-side effectively.

Is this something worth pursuing further? Is there any particular reason the two backends have to live apart? The underlying actions are the same, they're just triggered by different events. I'm sure they could be merged more gracefully, e.g. with options to enable them separately.

Just an idea, otherwise I'm extremely happy with the functionality provided by both backends.

**Comment by gaearon** [comment]
I think it is useful! I just don't have the time to work on something like this at all now.

**Comment by alexandercrichton** [comment]
I totally appreciate that. I know you're doing so much other amazing work.

Unfortunately I also need to keep moving. I was just thinking, since they've been combined here, could this be merged back into one of the main repo's? Then they could both be easily/optionally consumable, and could both be maintained in the one place. There are undoubtedly  further refactoring and optimisation opportunities that somebody with time could pursue, but for consumers this should add optional functionality with no drawbacks (that I can see).

After I'd tidied it up some more, would you object to a PR? What do you think @longlho, would you endorse this?

**Comment by longlho** [comment]
Sorry I'm on vacation and have limited access to email :( will try to take
a look next week.

**Comment by gaearon** [comment]
The problem with merging it back into the original repo is that I don’t have the time to maintain it. This will undoubtedly be useful but will also make new bugs appear, and I won’t be around to help fixing them. I think we should add an example to React DnD repo with this backend. Then it will be discoverable, and if enough people use it, maybe we can just start recommending it instead.

**Comment by alexandercrichton** [comment]
That sounds like a great idea. Who should do that? Is it an update that I can make?

**Comment by gaearon** [comment]
You can send a PR against React DnD. You have my approval in this thread so feel free to link back. Some fine folks are helping me out there so they’ll merge it.

**Comment by longlho** [comment]
@alexandercrichton yeah I'm down as long as it doesn't bloat up both backends. Tricky thing is hybrid devices that support both

**Comment by alexandercrichton** [comment]
@longlho that's actually the beauty of it. I've tried it on a surface and you can use mouse _and_ touch. I didn't even have to do anything, they both just work without affecting each other. The events they listen for don't overlap, so only one is actually active at a time.

It does make sense to keep the separation of concerns, but this is the only way I can see to support both types of input, unless... DragDropContext is allowed to accept multiple backends? What I see the DragDropContext doing is instantiating a DragDropManager which is passed the backend. I don't see why it couldn't instantiate multiple multiple managers with different backends. The managers appear to be self-contained such that they wouldn't affect each other. The backends are essentially event listeners, and as long as the consumer accepts responsibility for using appropriate backends they could have more than one listening at a time. I do see the other ramifications where various things expect to get just one manager.

Is there some definite reason that multiple backends isn't possible? Is it a matter of having time to implement it? or is it just not feasible/not a good idea?

**Comment by longlho** [comment]
@alexandercrichton  tl;dr: it's complexity, performance & maintainability but u should give it a try if ur interested :)

I think it's both complexity and performance, especially w/ `touch*` (or `mouse*`) events. It has to both listen & capture every related event for 1. Besides long tap has a different meaning in each browser, depending on the semantic element (`<a>` vs `<button>`...), and it also affects scrolling. Then it also may propagate to `mouse*` depending on whether u allow propagation, and which one u allow propagation. It's basically a guessing game. `drag*` is much more straightforward cause it indicates the intention already, but it's also more complex since it handles out of viewport drag in and all that jazz.

I think it can be done, I just don't know exactly the tradeoffs between bundling both together in 1 backend vs. letting the lib consumer decide at runtime which backend to use vs device capability. The touch backend right now is nowhere near feature-parity w/ the html5-backend and a lot of events don't happen for both so I think you might be looking at yet another abstraction layer that maps those events to a common set of operations to do.
